internal/service/common/repo/repository.go

```diff
@@ -84,3 +89,111 @@ func (r *CommonRepository) DeleteDataChangeEvent(ctx context.Context, id uuid.UU
+func ClaimDataChangeEvent(pool *pgxpool.Pool, ctx context.Context) ([]models2.DataChangeEvent, error) {
+	dataChangeEvents := make([]models2.DataChangeEvent, 0)
+	if err := pgx.BeginFunc(ctx, pool, func(tx pgx.Tx) error {
+		// Retrieve the current high watermark.
+		highWatermark, err := getHighWatermark(ctx, tx)
+		if err != nil {
+			return fmt.Errorf("failed to get high watermark: %w", err)
+		}
+
+		// Claim new outbox events (use FOR UPDATE SKIP LOCKED to avoid duplicate claims).
+		dataChangeEvents, err = getLatestDataChangeEvent(ctx, tx, highWatermark)
+		if err != nil {
+			return fmt.Errorf("failed to get latest data change events: %w", err)
+		}
+
+		// Update the high watermark if there are new events
+		if err := updateHighWatermark(ctx, tx, dataChangeEvents, highWatermark); err != nil {
+			return fmt.Errorf("failed to update high watermark: %w", err)
+		}
+
+		return nil
+	}); err != nil {
+		return dataChangeEvents, fmt.Errorf("failed transaction to claim notifications: %w", err)
+	}
+
+	return dataChangeEvents, nil
+}
+
+func getLatestDataChangeEvent(ctx context.Context, tx pgx.Tx, highWatermark int) ([]models2.DataChangeEvent, error) {
+	m := models2.DataChangeEvent{}
+	all := utils.GetAllDBTagsFromStruct(m)
+
+	queryGetLatestEvents := psql.Select(
+		sm.Columns(all.Columns()...),
+		sm.From(m.TableName()),
+		sm.Where(psql.Quote(all["SequenceID"]).GT(psql.Arg(highWatermark))),
+		sm.OrderBy(all["SequenceID"]).Asc(),
+		sm.ForUpdate(m.TableName()).SkipLocked(),
+	)
+
+	sql, params, err := queryGetLatestEvents.Build()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build queryGetLatestEvents query: %w", err)
+	}
+
+	dataChangeEvents, err := utils.ExecuteCollectRows[models2.DataChangeEvent](ctx, tx, sql, params)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute queryGetLatestEvents: %w", err)
+	}
+	return dataChangeEvents, nil
+}
+
+// getHighWatermark from notification_cursor get last_event_id to get the highwatermark
+func getHighWatermark(ctx context.Context, tx pgx.Tx) (int, error) {
+	m := models2.NotificationCursor{}
+	all := utils.GetAllDBTagsFromStruct(m)
+
+	queryGetHighWater := psql.Select(
+		sm.Columns(all["LastEventID"]),
+		sm.From(m.TableName()),
+		sm.Where(psql.Quote(all["ID"]).EQ(psql.Arg(1))),
+	)
+
+	sql, params, err := queryGetHighWater.Build()
+	if err != nil {
+		return 0, fmt.Errorf("failed to build queryGetHighWater query: %w", err)
+	}
+
+	highWatermark, err := utils.ExecuteCollectExactlyOneRow[models2.NotificationCursor](ctx, tx, sql, params)
+	if err != nil {
+		return 0, fmt.Errorf("failed to execute queryGetHighWater: %w", err)
+	}
+
+	return highWatermark.LastEventID, nil
+}
+
+// updateHighWatermark update notification cursor with the highest datachange event
+func updateHighWatermark(ctx context.Context, tx pgx.Tx, dataChangeEvents []models2.DataChangeEvent, highWatermark int) error {
+	if len(dataChangeEvents) > 0 {
```

Comment:
Could rewrite this to the following and save the nesting level.

```
if len(dataChangeEvents) == 0 {
    return nil
}
```

Reply:
updated